framework/src/meshgenerators/Poly2TriMeshGenerator.C

```diff
@@ -0,0 +1,378 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "Poly2TriMeshGenerator.h"
+
+#include "CastUniquePointer.h"
+#include "MooseMeshUtils.h"
+#include "MooseUtils.h"
+
+#include "libmesh/elem.h"
+#include "libmesh/int_range.h"
+#include "libmesh/mesh_modification.h"
+#include "libmesh/mesh_serializer.h"
+#include "libmesh/mesh_triangle_holes.h"
+#include "libmesh/parsed_function.h"
+#include "libmesh/poly2tri_triangulator.h"
+#include "libmesh/unstructured_mesh.h"
+
+registerMooseObject("MooseApp", Poly2TriMeshGenerator);
+
+InputParameters
+Poly2TriMeshGenerator::validParams()
+{
+  InputParameters params = MeshGenerator::validParams();
+
+  MooseEnum algorithm("BINARY EXHAUSTIVE", "BINARY");
+
+  params.addRequiredParam<MeshGeneratorName>(
+      "boundary", "The input MeshGenerator that defines the output mesh outer boundary.");
+  params.addParam<std::vector<BoundaryName>>(
+      "input_boundary_names", "2D-input-mesh boundaries defining the output mesh outer boundary");
+  params.addParam<std::vector<SubdomainName>>(
+      "input_subdomain_names", "1D-input-mesh subdomains defining the output mesh outer boundary");
+  params.addParam<unsigned int>("add_nodes_per_boundary_segment",
+                                0,
+                                "How many more nodes to add in each outer boundary segment.");
+  params.addParam<bool>(
+      "refine_boundary", true, "Whether to allow automatically refining the outer boundary.");
+
+  params.addParam<subdomain_id_type>(
+      "output_subdomain_id", 0, "Subdomain id to set on new triangles.");
+  params.addParam<SubdomainName>("output_subdomain_name",
+                                 "Subdomain name to set on new triangles.");
+
+  params.addParam<boundary_id_type>(
+      "output_boundary_id", 0, "Boundary id to set on new outer boundary.");
+  params.addParam<std::vector<boundary_id_type>>(
+      "hole_boundary_ids", "Boundary ids to set on holes.  Numbered up from 1 by default.");
```

Comment:
Why not `BoundaryName` and `SubdomainName` everywhere?

Reply:
`BoundaryName` options for `hole_boundary_ids` and `output_boundary_id` would be good, yeah.  What do you mean about `SubdomainName`?  Is there some way to combine `output_subdomain_id` and `output_subdomain_name` into one option?

Reply:
Yeah. Typically what we would do is if we parse a name for `output_subdomain`, we find a new ID and give it that name. If we parse an ID, we just set an ID